src/CodeGen_Hexagon.cpp

```diff
@@ -354,6 +306,64 @@ llvm::Value *CodeGen_Hexagon::callLLVMIntrinsic(llvm::Intrinsic::ID id,
     return callLLVMIntrinsic(Intrinsic::getDeclaration(module.get(), id), Ops);
 }
 
+namespace {
+Intrinsic::ID select_intrinsic(Type type,
+                               Intrinsic::ID b, Intrinsic::ID h, Intrinsic::ID w,
+                               Intrinsic::ID ub, Intrinsic::ID uh, Intrinsic::ID uw) {
+    switch (type.bits()) {
+    case 8: return type.is_int() ? b : ub;
+    case 16: return type.is_int() ? h : uh;
+    case 32: return type.is_int() ? w : uw;
+    }
+
+    internal_error << "Type not handled for intrinsic.\n";
+    return Intrinsic::not_intrinsic;
+}
+
+Intrinsic::ID select_intrinsic(Type type,
+                               Intrinsic::ID b, Intrinsic::ID h, Intrinsic::ID w) {
+    return select_intrinsic(type, b, h, w, b, h, w);
+}
+}  // namespace
+
+llvm::Value *CodeGen_Hexagon::call_intrin(Type t, int intrin_lanes,
+                                          llvm::Intrinsic::ID id,
+                                          int ops_lanes,
+                                          std::vector<Expr> ops) {
+    vector<Value *> values;
+    values.reserve(ops.size());
+    for (Expr i : ops) {
+        values.push_back(codegen(i));
+    }
+
+    return call_intrin(llvm_type_of(t), intrin_lanes, id, ops_lanes, values);
+}
+
+llvm::Value *CodeGen_Hexagon::call_intrin(llvm::Type *t, int intrin_lanes,
+                                          llvm::Intrinsic::ID id,
+                                          int ops_lanes,
+                                          std::vector<llvm::Value *> ops) {
+    internal_assert(id != Intrinsic::not_intrinsic);
+    llvm::Function *fn = Intrinsic::getDeclaration(module.get(), id);
+    llvm::FunctionType *fty = fn->getFunctionType();
+    internal_assert(fty->getNumParams() == ops.size());
+    for (size_t i = 0; i < ops.size(); i++) {
+        llvm::Type *ty = fty->getParamType(i);
+        if (ty != ops[i]->getType()) {
+            // We might need to cast, but the operands might have a
+            // different number of lanes. Fix up vector types before
+            // trying to cast.
+            if (ty->isVectorTy()) {
+                ty = llvm::VectorType::get(ty->getVectorElementType(),
+                                           ops_lanes*ty->getVectorNumElements()/intrin_lanes);
+            }
```

Comment:
I am not sure I understand the above. Don't both intrin_lanes and ty->getVectorNumElements() evaluate to the same thing since they are both related to the Intrinsic id?